infra/terraform/modules/data_access/memberships/memberships.py

```diff
@@ -52,10 +53,30 @@ def attach_bucket_policy(event, context):
     team_slug = event["team"]["slug"]
 
     client = boto3.client("iam")
-    client.attach_role_policy(
-        RoleName=naming.role_name(username),
-        PolicyArn=policy_arn(team_slug, policy_type),
-    )
+
+    # If role or policy are not there yet, retry
+    RETRY_DELAY = 0.200
+    tries = 2
+    while tries > 0:
+        try:
+            client.attach_role_policy(
+                RoleName=naming.role_name(username),
+                PolicyArn=policy_arn(team_slug, policy_type),
+            )
+            break
+        except botocore.exceptions.ClientError as error:
+            if error.response["Error"]["Code"] != "NoSuchEntity":
+                # Only retry on boto's ClientError/NoSuchEntity error
+                raise
+
+            tries -= 1
+            LOG.warning(
+                "error while attaching role to policy: {}.".format(error))
+            if tries > 0:
+                LOG.warning("Retry in {}s...".format(RETRY_DELAY))
+                time.sleep(RETRY_DELAY)
+            else:
+                raise
```

Comment:
I moved the retry logic in a function. I kept it simple and avoided weird things like annotations, etc...`retry` is just a function which takes the function to retry.